### ISSUE TITLE
logFromMCtoTG/logFromTGtoMC

### ISSUE THREAD

**Issue by AmidoNI2** [kraftwerk28/spigot-tg-bridge#1]
when the plugin needs to do logFromMCtoTG / logFromTGtoMC, the sending of the command / players online / time is normal, but not the messages from players, there are occasional moments that they receive from both sides, what do we do wrong?

and is there any way to change the / team? it conflicts with my Captcha Bot on Telegram

Minecraft 1.15.2

**Comment by kraftwerk28** [comment]
Thank you for posting the issue. What do you mean under moments when players receive messages from both sides?
Flag `logFromMCtoTG` means that messages that players type in the game chat will be sent by the bot in chats, which are specified in `chats` array in plugin's config. Flag `logFromTGtoMC` means that when bot receives any message from it's chats and if that chat is specified in plugin's config, it will be sent into the game chat.

Yes, I will try to make possible to set custom command text, thank you for submitting the idea.

**Comment by AmidoNI2** [comment]
**

> What do you mean under moments when players receive messages from both sides?

**

`I say that if inside the MC I send a message, it arrives at Telegram normally, but when messages from Telegram are sent to the MC, they rarely appear in the game`

**

> Flag logFromMCtoTG means that messages that players type in the game chat will be sent by the bot in chats, which are specified in chats array in plugin's config. Flag logFromTGtoMC means that when bot receives any message from it's chats and if that chat is specified in plugin's config, it will be sent into the game chat.

**

My Flags:
```
serverStartMessage: 'Server started.'
serverStopMessage: 'Server stopped.'
logJoinLeave: true
logFromMCtoTG: true
logFromTGtoMC: true
```
ThankYou for reply ^^

**Comment by ArtYurchenko** [comment]
> **
> 
> > What do you mean under moments when players receive messages from both sides?
> 
> **
> 
> `I say that if inside the MC I send a message, it arrives at Telegram normally, but when messages from Telegram are sent to the MC, they rarely appear in the game`
> 
> **
> 
> > Flag logFromMCtoTG means that messages that players type in the game chat will be sent by the bot in chats, which are specified in chats array in plugin's config. Flag logFromTGtoMC means that when bot receives any message from it's chats and if that chat is specified in plugin's config, it will be sent into the game chat.
> 
> **
> 
> My Flags:
> 
> ```
> serverStartMessage: 'Server started.'
> serverStopMessage: 'Server stopped.'
> logJoinLeave: true
> logFromMCtoTG: true
> logFromTGtoMC: true
> ```
> 
> ThankYou for reply ^^

What kernel of Minecraft do you use? I use this plugin on Spigot 1.15.2 and it works fine

**Comment by BlueWolfProduction** [comment]
Same issue here. It's hard to test for me on a live server, but talking to the telegram bot directly it seems to work. But as soon as I have it listen on a group chat it stops working properly. I send 10 messages, only 2 arrived. The other way (from mc to tg) works fine. /time and /online react similar.

I'm running Spigot 1.15.2.

**Comment by BlueWolfProduction** [comment]
I found something, but I can't tell if it's related or not. The MAIN issue I have is that it only occasionally reacts to telegram messages and commands. I have yet to find any patterns for it. But additionally, I found two separate crashes that makes it stop listening to telegram all together. I will just add them here.

The first one happened after I reloaded the plugin and send a /online command as my first message. This crash doesn't always happened. It happened after a fresh reload, until I did a full unload and load.

`[13:43:53] [mybot Telegram Executor/WARN]: Exception in thread "mybot Telegram Executor" 
[13:43:53] [mybot Telegram Executor/WARN]: java.lang.LinkageError: loader constraint violation: loader (instance of org/bukkit/plugin/java/PluginClassLoader) previously initiated loading for a different type with name "kotlin/jvm/functions/Function1"
[13:43:53] [mybot Telegram Executor/WARN]:     at java.lang.ClassLoader.defineClass1(Native Method)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:163)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.bukkit.plugin.java.JavaPluginLoader.getClassByName(JavaPluginLoader.java:202)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:107)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:96)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.lang.ClassLoader.defineClass1(Native Method)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.lang.ClassLoader.defineClass(ClassLoader.java:757)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:163)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.bukkit.plugin.java.JavaPluginLoader.getClassByName(JavaPluginLoader.java:202)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:107)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.bukkit.plugin.java.PluginClassLoader.findClass(PluginClassLoader.java:96)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.lang.ClassLoader.loadClass(ClassLoader.java:419)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.lang.ClassLoader.loadClass(ClassLoader.java:352)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.kraftwerk28.spigot_tg_bridge.Bot.onUpdateReceived(Bot.kt:47)
[13:43:53] [mybot Telegram Executor/WARN]:     at java.util.ArrayList.forEach(ArrayList.java:1257)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.telegram.telegrambots.meta.generics.LongPollingBot.onUpdatesReceived(LongPollingBot.java:27)
[13:43:53] [mybot Telegram Executor/WARN]:     at org.telegram.telegrambots.updatesreceivers.DefaultBotSession$HandlerThread.run(DefaultBotSession.java:312)`

The second message is similar, but happened when I did either send /online or /time as my first message.

`[13:46:16] [mybot Telegram Executor/WARN]: Exception in thread "mybot Telegram Executor" 
[13:46:16] [mybot Telegram Executor/WARN]: java.lang.IllegalAccessError: tried to access class org.kraftwerk28.spigot_tg_bridge.Bot$onUpdateReceived$2 from class org.kraftwerk28.spigot_tg_bridge.Bot
[13:46:16] [mybot Telegram Executor/WARN]:     at org.kraftwerk28.spigot_tg_bridge.Bot.onUpdateReceived(Bot.kt:59)
[13:46:16] [mybot Telegram Executor/WARN]:     at java.util.ArrayList.forEach(ArrayList.java:1257)
[13:46:16] [mybot Telegram Executor/WARN]:     at org.telegram.telegrambots.meta.generics.LongPollingBot.onUpdatesReceived(LongPollingBot.java:27)
[13:46:16] [mybot Telegram Executor/WARN]:     at org.telegram.telegrambots.updatesreceivers.DefaultBotSession$HandlerThread.run(DefaultBotSession.java:312)`

Note that I changed my bot identity.

I'm running Spigot 1.15.2. Just updated actually.

**Comment by kraftwerk28** [comment]
Hey, I think I figured out why that happens in you case. Try to enable groups in your bot settings.
![image](https://user-images.githubusercontent.com/31807671/82714249-72024400-9c96-11ea-8365-62653e4acc7c.png)

**Comment by AmidoNI2** [comment]
> 
> 
> > **
> > > What do you mean under moments when players receive messages from both sides?
> > 
> > 
> > **
> > `I say that if inside the MC I send a message, it arrives at Telegram normally, but when messages from Telegram are sent to the MC, they rarely appear in the game`
> > **
> > > Flag logFromMCtoTG means that messages that players type in the game chat will be sent by the bot in chats, which are specified in chats array in plugin's config. Flag logFromTGtoMC means that when bot receives any message from it's chats and if that chat is specified in plugin's config, it will be sent into the game chat.
> > 
> > 
> > **
> > My Flags:
> > ```
> > serverStartMessage: 'Server started.'
> > serverStopMessage: 'Server stopped.'
> > logJoinLeave: true
> > logFromMCtoTG: true
> > logFromTGtoMC: true
> > ```
> > 
> > 
> > ThankYou for reply ^^
> 
> What kernel of Minecraft do you use? I use this plugin on Spigot 1.15.2 and it works fine

spigot with 1.15.2

**Comment by kraftwerk28** [comment]
@AmidoNI2 did you try to change bot settings in Telegram?

**Comment by BlueWolfProduction** [comment]
> Maybe this happens due to incorrect bot settings. Try to enable groups through BotFather.
> ![image](https://user-images.githubusercontent.com/31807671/82714249-72024400-9c96-11ea-8365-62653e4acc7c.png)

Excuse me for my delayed reply. I can confirm groups were turned on, and privacy mode was turned off. I previously used that account for my own project in groups and never had problems with it.

Now... It has started to work out of nowhere. I have repeatedly disabled/enabled the plugin and repeatedly restarted the entire server. And for a while it did not work properly but I decided to leave the plugin loaded. I've been working on my own (not related) plugin that required me to restart the server some more times until I realised this plugin has started to work out of nowhere.

It has been a few days, and has continued working even when I restarted the server some more times. I don't know what happened, nothing has changed. It just started working.

Now I did try other telegram plugins. And I did unload them before I moved on to the next. But now I'm wondering if it was competing with other (not properly unloaded) plugins, or plugins that accidentally started again after I rebooted the server. I did change the bot token, but then I started having the /command errors so... I'm not sure though.

So I don't know what made it work. But I do have a suggestion for others who are having the problem. Did you try any other telegram plugins before this? Have they all shut down properly? If you load plugins with another plugin, it might be worth completely restarting the server to be completely sure. The way it acted makes me think two programs are competing with the same bot. Consider changing your bot token to kick out any rough plugins still loaded in memory.
